txmempool: split epoch logic into class

ajtowns opened the pull request:
Splits the epoch logic introduced in #17925 into a separate class.

Uses clang's thread safety annotations and encapsulates the data more strongly to reduce chances of bugs from API misuse.

JeremyRubin commented:
Concept ACK. Will need to play around with it a little bit to test that clang actually prevents compiling incorrect uses.

laanwj commented:
Concept ACK, nice

hebasto commented:
Concept ACK.

hebasto commented on src/txmempool.h:
```diff
@@ -50,6 +50,68 @@ struct LockPoints
+ * algorithm.
+ */
+
+class Epoch
```
Tested on top of the #18635, and had loads of warnings:
```
./txmempool.h:102:38: warning: 'exclusive_lock_function' attribute requires arguments whose type is annotated with 'capability' attribute; type here is 'Epoch' [-Wthread-safety-attributes]
        explicit Guard(Epoch& epoch) EXCLUSIVE_LOCK_FUNCTION(epoch) LOCKS_EXCLUDED(epoch);
                                     ^
...
./txmempool.h:106:40: warning: 'exclusive_locks_required' attribute requires arguments whose type is annotated with 'capability' attribute; type here is 'const Epoch' [-Wthread-safety-attributes]
    bool visited(Marker& marker) const EXCLUSIVE_LOCKS_REQUIRED(*this) {
                                       ^
...
./txmempool.h:691:121: warning: 'locks_excluded' attribute requires arguments whose type is annotated with 'capability' attribute; type here is 'Epoch' [-Wthread-safety-attributes]
    void UpdateTransactionsFromBlock(const std::vector<uint256>& vHashesToUpdate) EXCLUSIVE_LOCKS_REQUIRED(cs, cs_main) LOCKS_EXCLUDED(m_epoch);
                                                                                                                        ^
...
./txmempool.h:812:41: warning: 'exclusive_locks_required' attribute requires arguments whose type is annotated with 'capability' attribute; type here is 'Epoch' [-Wthread-safety-attributes]
    bool visited(const txiter it) const EXCLUSIVE_LOCKS_REQUIRED(cs, m_epoch) {
                                        ^
...
./txmempool.h:816:45: warning: 'exclusive_locks_required' attribute requires arguments whose type is annotated with 'capability' attribute; type here is 'Epoch' [-Wthread-safety-attributes]
    bool visited(Optional<txiter> it) const EXCLUSIVE_LOCKS_REQUIRED(cs, m_epoch) {
                                            ^
```

To make the Clang Thread Safety Analysis work as expected I suggest: 
```suggestion
class LOCKABLE Epoch
```

_and_ change the type of the `Epoch::Guard::m_epoch` from `Epoch&` to `Epoch*`.

ajtowns replied:
I'm not seeing what benefit changing `Epoch::Guard::m_epoch` from reference to pointer would have -- the analysis shouldn't (and doesn't) treat them any different, as far as I can see?

hebasto replied:
Indeed :)
IDK why I saw warnings...

hebasto commented on src/txmempool.h:
```diff
@@ -50,6 +50,68 @@ struct LockPoints
+        Epoch& m_epoch;
+
+    public:
+        explicit Guard(Epoch& epoch) EXCLUSIVE_LOCK_FUNCTION(epoch) LOCKS_EXCLUDED(epoch);
```
From the Thread Safety Analysis [docs](https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#acquire-acquire-shared-release-release-shared):
> The caller must not hold the given capability on entry

It seems `LOCKS_EXCLUDED(epoch)` is redundant, no?

hebasto reviewed (changes requested):
A brilliant idea to leverage the Clang Thread Safety Analysis!

ajtowns commented:
Incorporated @hebasto's suggested changes

hebasto reviewed (approved):
ACK 837d4e4f5aeb2f110143c59b7dd05eac6ae52b63, tested on Linux Mint 19.3: the Clang's thread safety annotations indeed reduce chances of the `Epoch` class misuse (verified by different code manipulations).

May I suggest two additional annotations:
```diff
--- a/src/txmempool.cpp
+++ b/src/txmempool.cpp
@@ -107,6 +107,7 @@ void CTxMemPool::UpdateForDescendants(txiter updateIt, cacheMap &cachedDescendan
 void CTxMemPool::UpdateTransactionsFromBlock(const std::vector<uint256> &vHashesToUpdate)
 {
     AssertLockHeld(cs);
+    AssertLockNotHeld(m_epoch);
     // For each entry in vHashesToUpdate, store the set of in-mempool, but not
     // in-vHashesToUpdate transactions, so that we don't have to recalculate
     // descendants when we come across a previously seen entry.
diff --git a/src/txmempool.h b/src/txmempool.h
index 655afc3b8..b72ad229c 100644
--- a/src/txmempool.h
+++ b/src/txmempool.h
@@ -797,7 +797,7 @@ private:
      */
     void UpdateForDescendants(txiter updateIt,
             cacheMap &cachedDescendants,
-            const std::set<uint256> &setExclude) EXCLUSIVE_LOCKS_REQUIRED(cs);
+            const std::set<uint256> &setExclude) EXCLUSIVE_LOCKS_REQUIRED(cs) LOCKS_EXCLUDED(m_epoch);
     /** Update ancestors of hash to add/remove it as a descendant transaction. */
     void UpdateAncestorsOf(bool add, txiter hash, setEntries &setAncestors) EXCLUSIVE_LOCKS_REQUIRED(cs);
     /** Set ancestor state for an entry */
```
?

ajtowns commented:
@hebasto - `AssertLockNotHeld(m_epoch)` won't work -- epochs aren't sync.h locks. I think the annotation for UpdateForDescendents would be better made after #18191 is merged?

hebasto commented:
@ajtowns 
> @hebasto - `AssertLockNotHeld(m_epoch)` won't work -- epochs aren't sync.h locks. I think the annotation for UpdateForDescendents would be better made after #18191 is merged?

Agree.

ajtowns commented:
Rebased

JeremyRubin reviewed (approved):
Code Review ACK.

ajtowns commented on src/txmempool.h:
```diff
@@ -63,6 +63,64 @@ struct CompareIteratorByHash {
+    Epoch(const Epoch&) = delete;
+    Epoch& operator=(const Epoch&) = delete;
+
+    bool guarded() const { return m_guarded; }
```
There's an `assert(m_epoch.guarded());` in `CTxMemPool::visited(Optional<txiter> it)` to check the lock's held in the case where `it->visited` isn't called.

ajtowns commented on src/txmempool.h:
```diff
@@ -63,6 +63,64 @@ struct CompareIteratorByHash {
+        bool ret = marker.m_marker >= m_raw_epoch;
+        if (!ret) marker.m_marker = m_raw_epoch;
+        return ret;
+    }
```
That reverses the logic (`ret == false` causes `m_marker` to change originally, but your code returns `true` if `m_marker` changed) ? Seems like that's evidence it's not that easy to read/reason about? :)

It might be less confusing to write it out in full:

```c++
if (marker.m_marker < m_raw_epoch ) {
    // this entry is from an earlier epoch, so it hasn't been visited
    marker.m_marker = m_raw_epoch;
    return false;
} else {
    return true;
}
```

I'd guess the compiler will make all these variants essentially the same anyway, so no objection to any variation from me.

JeremyRubin replied:
Ah my bad, I've made a case against myself -- for what it's worth, I made the proposed version by translating the proposed version, which I have trouble parsing logically, and thought it was equivalent to that logic (and still can't figure out the concrete reason it's opposite).

I'm fine with the completely written out version as it is the most obvious for sleepy brains like mine.

ajtowns replied:
I think the translation goes:

- `r = a >= b; if (!r) a = b; return r;`  (current)
- `o = a; if (! (a >= b)) a = b; return o >= b;`   (store old value, replace `r`)
- `o = a; if (a < b) a = b; return !(o < b);`  (switch to less than)
- `o = a; a = max(a,b); return !(o != a);`  (`o < b` iff the if was taken, and `a` was changed)
- `o = a; a = max(a,b); return o == a;`  (simplify)

Anyway, changed to KISS version.

ajtowns commented on src/txmempool.cpp:
```diff
@@ -1113,22 +1113,17 @@ void CTxMemPool::SetIsLoaded(bool loaded)
 
 
-CTxMemPool::EpochGuard CTxMemPool::GetFreshEpoch() const
+Epoch::Guard::Guard(Epoch& epoch) : m_epoch(epoch)
```
Moved into util/epochguard.h

JeremyRubin commented on src/txmempool.h:
```diff
@@ -814,52 +815,20 @@ class CTxMemPool
-    bool visited(Optional<txiter> it) const EXCLUSIVE_LOCKS_REQUIRED(cs) {
-        assert(m_has_epoch_guard);
+    bool visited(Optional<txiter> it) const EXCLUSIVE_LOCKS_REQUIRED(cs, m_epoch) {
+        assert(m_epoch.guarded());
```
micronit: may be worth adding a comment here that the assert is redundant when it != nullopt, but is used for consistency when it == nullopt.

JeremyRubin commented:
re CR-ACK

jonatack commented:
Concept ACK, the code looks like a nice refactoring.

ajtowns committed:
[refactor] txmempool: split epoch logic into class

ajtowns commented:
Rebased to deal with adjacent lines changed in #19935

felipsoarez reviewed (commented):
utACK

hebasto commented on src/txmempool.h:
```diff
@@ -21,6 +21,7 @@
 #include <primitives/transaction.h>
 #include <sync.h>
 #include <random.h>
+#include <util/epochguard.h>
```
style nit: while touching headers they could be sorted:
```
#include <random.h>
#include <sync.h>
#include <util/epochguard.h>
```

ajtowns replied:
Sorted in #20944

hebasto commented on src/txmempool.h:
```diff
@@ -819,52 +820,20 @@ class CTxMemPool
-        bool ret = it->m_epoch >= m_epoch;
-        it->m_epoch = std::max(it->m_epoch, m_epoch);
-        return ret;
+    bool visited(const txiter it) const EXCLUSIVE_LOCKS_REQUIRED(cs, m_epoch) {
```
style nit, suggested by `clang-format`:
```suggestion
    bool visited(const txiter it) const EXCLUSIVE_LOCKS_REQUIRED(cs, m_epoch)
    {
```

hebasto commented on src/txmempool.h:
```diff
@@ -819,52 +820,20 @@ class CTxMemPool
 
-    bool visited(Optional<txiter> it) const EXCLUSIVE_LOCKS_REQUIRED(cs) {
-        assert(m_has_epoch_guard);
+    bool visited(Optional<txiter> it) const EXCLUSIVE_LOCKS_REQUIRED(cs, m_epoch) {
```
style nit, suggested by `clang-format`:
```suggestion
    bool visited(Optional<txiter> it) const EXCLUSIVE_LOCKS_REQUIRED(cs, m_epoch)
    {
```

hebasto commented on src/util/epochguard.h:
```diff
@@ -0,0 +1,88 @@
+    bool m_guarded = false;
+
+public:
+    Epoch() { }
```
style nit, suggested by `clang-format`:
```suggestion
    Epoch() {}
```

hebasto commented on src/util/epochguard.h:
```diff
@@ -0,0 +1,88 @@
+
+    bool guarded() const { return m_guarded; }
+
+    class Marker {
```
style nit, suggested by `clang-format`:
```suggestion
    class Marker
    {
```

hebasto commented on src/util/epochguard.h:
```diff
@@ -0,0 +1,88 @@
+        Marker& operator=(const Marker&) = delete;
+    };
+
+    class SCOPED_LOCKABLE Guard {
```
style nit, suggested by `clang-format`:
```suggestion
    class SCOPED_LOCKABLE Guard
    {
```

hebasto commented on src/util/epochguard.h:
```diff
@@ -0,0 +1,88 @@
+ * visiting the same transactions twice. Some traversal algorithms use
+ * std::set (or setEntries) to deduplicate the transaction we visit.
+ * However, use of std::set is algorithmically undesirable because it both
+ * adds an asymptotic factor of O(log n) to traverals cost and triggers O(n)
```
typo, https://github.com/bitcoin/bitcoin/pull/17925#discussion_r374052497:
```suggestion
 * adds an asymptotic factor of O(log n) to traversals cost and triggers O(n)
```

hebasto commented on src/util/epochguard.h:
```diff
@@ -0,0 +1,88 @@
+            m_epoch.m_guarded = true;
+        }
+        ~Guard() UNLOCK_FUNCTION()
+        {
+            ++m_epoch.m_raw_epoch; // ensure clear separation between epochs
```
```suggestion
        {
            assert(m_epoch.m_guarded);
            ++m_epoch.m_raw_epoch; // ensure clear separation between epochs
```

hebasto reviewed (commented):
ACK 88019f9a183d396713fd357604a6e472c5ed8807

It seems both `CTxMemPool::visited` member functions could be `private`.

ajtowns commented:
:birthday: This PR turned 1 the other week! Rebased on top of #20944 due to header reordering, and addressed @hebasto's nits.

(I think the mempool visited methods are public deliberately, so that they could be used by external functions if desired; so haven't made them private)

JeremyRubin commented:
Yes this is correct; IIRC there's some stalled out patches which require them to be visible :)

jonatack reviewed (approved):
ACK fd6580e405699ccb051fd2a34525e48d3253673d using clang thread safety annotations looks like a very good idea, and the encapsulation this change adds should improve robustness (and possible unit test-ability) of the code. Verified that changing some of the locking duly provoked build-time warnings with Clang 9 on Debian and that small changes in the new `Epoch` class were covered by failing functional test assertions in `mempool_updatefromblock.py`, `mempool_resurrect.py`, and `mempool_reorg.py`

example clang warning
```
/txmempool.h:840:24: warning: calling function 'visited' requires holding mutex 'm_epoch' exclusively [-Wthread-safety-analysis]
        return m_epoch.visited(it->m_epoch_marker);
```

Ignore the two comments below unless you need to retouch anything (the documentation was move-only anyway).

hebasto reviewed (approved):
re-ACK fd6580e405699ccb051fd2a34525e48d3253673d, since my [previous](https://github.com/bitcoin/bitcoin/pull/18017#pullrequestreview-569619362) review:
- rebased
- style nits and typo addressed
- added `assert(m_epoch.m_guarded);` as [suggested](https://github.com/bitcoin/bitcoin/pull/18017#discussion_r558607064)
- the `Epoch` ctor is `default` now (was empty)
